earthaccess/search.py

```diff
@@ -11,6 +11,33 @@
 from .results import DataCollection, DataGranule
 
 
+def _normalize_datetime(raw: None | str | dt.date | dt.datetime) -> None | dt.datetime:
+    # empty string or None is None
+    if not raw:
+        return None
+    # the cmr.*Query.temporal method only parses a subset of ISO 8601 strings, so we
+    # add flexibility, inclusive of the presence or absence of timezone information
+    if isinstance(raw, str):
+        normalized = dateutil.parser.parse(raw)
+    elif not isinstance(raw, dt.datetime):
+        normalized = dt.datetime.combine(raw, dt.time())
```

Comment:
Or just wrap the whole bit and catch both type and attribute errors

Reply:
I think if (as in latest) we do the `dt.datetime.combine` only on `dt.date` objects (and not `dt.datetime` objects), then the subsequent try...catch will be sufficient?

Reply:
@itcarroll I just pushed a possibly overly comprehensive and explicit set of tests for this function, with what I would expect `_normalize_datetime` to do; can you look them over and see if you agree?

Also, *a lot* of the complexity can actually be avoided by relying more on `dateutil.parser.parse` with a default `datetime` like this:
```python
def _normalize_datetime(raw: None | str | dt.date | dt.datetime) -> None | dt.datetime:
    """Normalize datetime like objects for CMR

    Date-times provided to CMR are required to be in UTC and pyCMR's cmr.*Query.temporal method only parses a subset of
    ISO 8601 strings. This method allows us to support a wider variety of date-time-like objects, such as partial
    strings, dates, and datetime objects.

    Args:
        raw: An optional date-time-like object to be normalized.

    Returns:
        A datetime in UTC.
    """
    # empty string or None is None
    if not (normalized := raw):
        return None

    # switch to iso formatted string. Note: this also works for pandas timestamps but not numpy datetime64.
    if isinstance(raw, (dt.date, dt.datetime)):
        normalized = raw.isoformat()

    # parse date-time strings, rounding partial dates down and assuming UTC if timezone isn't provided
    try:
        normalized = dateutil.parser.parse(normalized, default=dt.datetime(1, 1, 1, 0, 0, tzinfo=dt.timezone.utc))
    except TypeError:
        raise TypeError(f"Dates must be date, datetime, or str objects, not  a {raw.__class__.__name__} object.")

    return normalized.astimezone(dt.timezone.utc)
```

What do you think?

Reply:
Thanks for all the tests. I hope I made it easier to maintain by converting the test with many asserts to parameterized test.

What I don't love about the simplification here is the call to parse when it may not have been necessary (if we already had a datetime). I didn't benchmark against my latest, but it does not round-trip a dt.datetime through a string.